### DISCUSSION TITLE
confused on how to use the limiter for 2 different requests but both need to use a limiter.

### DISCUSSION THREAD

**Discussion by outgaze** [mjpieters/aiolimiter#79] [Q&A] [answered]
Basically what i'm trying to do is

```py
if limiter.has_capacity(): # when this hits capacity go to limiter2 and do the request #2 within the limit.
 do request #1
elif limiter2.has_capacity():
 do request #2
```

the problem is i don't understand how to do it

**Comment by mjpieters** [comment, marked as answer]
Why not just put both requests into separate tasks? You are using asyncio after all and if you use two separate limiters for two separate URLs just run those in parallel.

**Reply by outgaze** [reply]
it's a request to the same api with different api keys but same input (uuids) each key has a limit of 110 requests/min, do you get it now? what i'm trying to do...

**Reply by outgaze** [reply]
like i need the data to be returned as one...

**Reply by outgaze** [reply]
basically if one limiter hits the limit it stops using the key to send the reqs on that key and continues with another limiter with the same amount, once the first one has free space it goes back to it..

**Reply by mjpieters** [reply]
This is more complex than it may appear on the surface, and using two limiters would be the wrong approach here. Instead, I'ad use a single limiter set to limit the rate to 220 requests per minute and simply alternate between the API keys.

Note also that the limiter shows [bursting behaviour](https://aiolimiter.readthedocs.io/en/latest/#bursting) that may be surprising when combined with rate-limited API keys. If you send 120 tasks to a limiter  set up with `AsyncLimiter(110)`, then 110 will execute immediately, and the remaining 10 will execute every 60 / 110 ≈ 0.54 seconds, meaning that you'll have sent 120 requests in under 6 seconds.

Instead, create a limiter that allows for a new request every 60 / 220 seconds, and use an [`itertools.cycle()` object](https://docs.python.org/3/library/itertools.html#itertools.cycle) to let you alternate the key used:

```python
limiter = AsyncLimiter(1, 60 / 220)
api_keys = cycle([API_KEY_1, API_KEY_2])

```python
async with limiter:
    # section limited to 220 req / minute
   key = next(api_key)  # uses either API_KEY_1 or API_KEY_2, whichever was not used for the previous request
```

This uses `API_KEY_1` at most 110 times per minute, and uses `API_KEY_2` at most 110 times per minute. If you send 220 requests to this task, the first request immediately runs using `API_KEY_1`, the second has to wait ~ 0.273 seconds and is sent using `API_KEY_2`, the third is allowed to enter the section after ~ 0.54 seconds and is sent using `API_KEY_1`, etc.

**Reply by outgaze** [reply]
Thank you a lot, you're a genius i'm new to python this really helped, you're amazing ill leave a star for sure on the repo.

**Reply by outgaze** [reply]
> This is more complex than it may appear on the surface, and using two limiters would be the wrong approach here. Instead, I'ad use a single limiter set to limit the rate to 220 requests per minute and simply alternate between the API keys.
> 
> Note also that the limiter shows [bursting behaviour](https://aiolimiter.readthedocs.io/en/latest/#bursting) that may be surprising when combined with rate-limited API keys. If you send 120 tasks to a limiter set up with `AsyncLimiter(110)`, then 110 will execute immediately, and the remaining 10 will execute every 60 / 110 ≈ 0.54 seconds, meaning that you'll have sent 120 requests in under 6 seconds.
> 
> Instead, create a limiter that allows for a new request every 60 / 220 seconds, and use an [`itertools.cycle()` object](https://docs.python.org/3/library/itertools.html#itertools.cycle) to let you alternate the key used:
> 
> ```python
> limiter = AsyncLimiter(1, 60 / 220)
> api_keys = cycle([API_KEY_1, API_KEY_2])
> 
> ```python
> async with limiter:
>     # section limited to 220 req / minute
>    key = next(api_key)  # uses either API_KEY_1 or API_KEY_2, whichever was not used for the previous request
> ```
> 
> This uses `API_KEY_1` at most 110 times per minute, and uses `API_KEY_2` at most 110 times per minute. If you send 220 requests to this task, the first request immediately runs using `API_KEY_1`, the second has to wait ~ 0.273 seconds and is sent using `API_KEY_2`, the third is allowed to enter the section after ~ 0.54 seconds and is sent using `API_KEY_1`, etc.

oh yeah also i forgot to ask, is there a way to make it do more requests as im adding more api keys for example let's say 5 api keys = 550 reqs/min and make it do more than 1 in order to be faster?
